developer_manual/digging_deeper/speech-to-text.rst

```diff
@@ -0,0 +1,138 @@
+    use OCP\SpeechToText\Events\TranscriptionFailedEvent;
+    use OCP\EventDispatcher\Event;
+    use OCP\EventDispatcher\IEventListener;
+
+    class MyTranscriptionListener implements IEventListener {
+        public function handle(Event $event): void {
+            if (!$event instanceof AbstractTranscriptionEvent || $event->getAppId() !== Application::APP_ID) {
+                return;
+            }
+
+            if ($event instanceof TranscriptionSuccessfulEvent) {
+                $transcript = $event->getTranscript()
+                // store $transcript somewhere
+            }
+
+            if ($event instanceof TranscriptionSuccessfulEvent) {
+                $error = $event->getErrorMessage()
+                $userId = $event->getUserId()
+                // Notify relevant user about failure
+            }
+        }
+    }
+
+
+Implementing a Speech-To-Text provider
+--------------------------------------
+
+A **Speech-To-Text provider** is a class that implements the interface ``OCP\SpeechToText\ISpeechToTextProvider``.
+
+.. code-block:: php
+
+    <?php
+
+    declare(strict_types=1);
+
+    namespace OCA\MyApp\SpeechToText;
+
+    use OCA\MyApp\AppInfo\Application;
+    use OCP\Files\File;
+    use OCP\SpeechToText\ISpeechToTextProvider;
+    use OCP\IL10N;
+
+    class Provider implements ISpeechToTextProvider {
+
+        public function __construct(
+            private IL10N $l,
+        ) {
+        }
+
+        public function getName(): string {
+            return $this->l->t('My awesome speech to text provider');
+        }
+
+        public function transcribeFile(File $file): string {
+            // transcribe file here and return transcript
+        }
+    }
+
+The method ``getName`` returns a string to identify the registered provider in the user interface.
+
+The method ``transcribeFile`` transcribes the passed file and returns the transcript. In case transcription fails, you should throw a ``RuntimeException`` with an explanatory error message.
+
+The class would typically be saved into a file in ``lib/SpeechToText`` of your app but you are free to put it elsewhere as long as it's loadable by Nextcloud's :ref:`dependency injection container<dependency-injection>`.
+
+Provider registration
+---------------------
+
+The provider class is registered via the :ref:`bootstrap mechanism<Bootstrapping>` of the ``Application`` class.
+
+.. code-block:: php
+
+
+    <?php
+
+    declare(strict_types=1);
+
+    namespace OCA\MyApp\AppInfo;
+
+    use OCA\MyApp\SpeechToText\Provider;
+    use OCP\AppFramework\App;
+    use OCP\AppFramework\Bootstrap\IBootContext;
+    use OCP\AppFramework\Bootstrap\IBootstrap;
+    use OCP\AppFramework\Bootstrap\IRegistrationContext;
+
+    class Application extends App implements IBootstrap {
+
+        public function register(IRegistrationContext $context): void {
+            $context->registerSpeechToTextProvider(Provider::class);
```

Comment:
highlight this line with `:emphasize-lines:` https://docs.nextcloud.com/server/latest/developer_manual/basics/middlewares.html#global-middlewares

Reply:
TIL, nice!